Texture filtering support, font rendering and some fixes

digetx opened the pull request:
Add texture filtering support to grate. The font rendering is used by a new texture-filter grate test.

digetx committed:
libhost1x: add host1x_bo_wrap()

It will be used by a texture mipmap generator, helps to keep code clean.

digetx committed:
grate: fragment asm/disasm: TEX instruction opcodes TEX/TXB

The TEX instruction now takes opcode to differentiate between LOD biased and
un-biased texture sampling. We already implement the TEX opcode, so the opcode is
just prepended to the instruction. TXB is the new opcode, it takes additional last
source-register, the LOD bias.

EXEC
	TEX:	tex r2, r3, tex14, r0, r1, r2
;

EXEC
	TEX:	txb r0, r1, tex3, r0, r1, r2, r3
;

digetx committed:
libhost1x: handle BO offset in GR2D and pixelbuffer

So the wrapped BO's are working correctly.

digetx committed:
libhost1x: host1x_gr2d_surface_blit() fixes and feature

HW is now set up properly, the feature-hack is scale of RGBA8888 format support.

digetx committed:
grate: remove unneeded 'static' declarations

Probably they were cut-n-pasted by mistake.

digetx committed:
grate: asm: destroy lexers

This resets lexers into initial state and frees previously allocated memory,
so fixes memleak.

digetx committed:
grate: asm: fix cgc symbols uniform name

The name must be copied from the parsed result. It happened that we haven't
used two shader programs with different symbol names in one application for now.

digetx committed:
grate: correct 3d context uniforms location check

This fixes check for ~0u value which we are getting for a requested uniform that
wasn't found.

digetx committed:
grate: mipmap texture support

Both autogeneration and manual load of LOD levels are implemented. The autogen
uses GR2D for texture scaling.

digetx committed:
grate: display text using 3d

It should be handy to display some textual information for the grate tests.
The font texture was generated by "Ultimate bitmap font generator" tool.

digetx committed:
tests/grate: add texture-filter test

Test minify/magnify/mipmap texture filtering modes.

kusma commented on src/libgrate/fragment_asm_parser.y:
```diff
@@ -765,6 +767,19 @@ TEX_INSTRUCTION:
+	|
+	{
+		asm_tex_instructions[asm_fs_instructions_nb].enable_bias = 0;
+		yyval.u = 0xDEAD;
```
The "common" way of doing this, is to consider these two different opcodes, [TEX](http://gallium.readthedocs.io/en/latest/tgsi.html#opcode-TEX) and [TXB](http://gallium.readthedocs.io/en/latest/tgsi.html#opcode-TXB)...

digetx replied:
Indeed, differentiating instructions by opcode should be more expressive and there will be more opcodes in the future. We should probably omit the 3-rd cubemap coord source register from the current TEX instruction and use it in the TEX2 instruction opcode (following TGSI ISA) once will add support for cubemaps. 

 I'll re-spin the series.

kusma replied:
TEX2 is about shadow cube-map arrays, though, how is that related?

digetx replied:
Ah, okay :) I just quickly glanced through the TGSI ISA without paying much attention. So the TEX opcode is fine as it is now, no need to change anything about it.

digetx commented:
Re-spun, now with the TXB opcode. Please take a look =)

kusma commented on src/libgrate/grate-texture.c:
```diff
@@ -184,3 +264,212 @@ void grate_texture_clear(struct grate *grate, struct grate_texture *tex,
+		return 1;
+	case 0 ... 1:
+		return 0;
+	}
```
This is a huge jump-table, using compiler-specific range-based case-statements... How about just doing __builtin_clz() instead? :)

digetx replied:
That's not true. I checked disassembly and the generated code is tiny, actually it is compiled into

if (size > 1024) {
} else if (size > 512) {
} ...

kusma replied:
Oh, OK. I still think a single instruction is slightly better, but it's not at all important. Let's just leave it like this for now.

digetx replied:
clz would also work, but I like more expressive approach. It would be insane if any relatively modern compiler won't optimize that switch-case. And again, the worst performance bottle neck is memory management, BO allocation time in particular.

kusma replied:
Fair enough!

digetx replied:
Yeah, one obvious optimization would be getting rid of CMA allocation on each command submit in the kernel driver. That gave a noticeable boost in performance while I was tackling the opentegra driver, actually not so long time ago. Also the current in-kernel command stream validation code completely kills performance, I usually have it turned off while hacking-developing.

kusma commented on src/libgrate/grate-3d.c:
```diff
@@ -659,11 +659,13 @@ static void grate_3d_set_texture_desc(struct host1x_pushbuf *pb,
+				      bool mipmap_enabled,
+				      bool min_filter_linear_within,
+				      bool min_filter_linear_between,
+				      bool mag_filter_linear)
```
This isn't really important, but I'm having a bit of difficulty liking this name-change. I know that I merged it in envytools, but I'm having a bit of second thought here... What's wrong with `mag`, `min`, `mip` names? These are the names that D3D uses for the filtering modes, I've never seen OpenGL refer to these apart from that wiki-site you listed (which is more of a textual explanation than "terminology")...

Basically, every time I read these, I get a little bit confused. I guess I'm just so used to the old names...

digetx replied:
I'm not used to any naming convention, so do not really care. Just tell me how to name them.

MIPFILTER == MINFILTER_LINEAR_WITHIN ?
MINFILTER == MINFILTER_LINEAR_BETWEEN ?

Is that ^ correct?

kusma replied:
The "min"-filter is what happens to a single texture image when it's being minified (similar to magfilter), and the "mip"-filter is how the results of the min/mag-filters gets combined. So unless I'm confused, I think that would be:

MINFILTER == MINFILTER_LINEAR_WITHIN
MIPFILTER == MINFILTER_LINEAR_BETWEEN

digetx replied:
Okay, I'll switch to these names. Let me know if there is anything else confusing :)

kusma replied:
@digetx: Nah, I think it all looks very reasonable. Sorry for being a bit nit-picky here ;)

digetx replied:
Since I'm not very familiar with OpenGL and related terminology in general, I'm expecting that from you :) Don't shy, always let me know if something feels clumsy.

digetx committed:
grate: use booleans for the texture warp modes

That is a bit nicer than a bitmask.

digetx commented:
Changed the names, please take a look. 🔍 :)

kusma commented on src/libgrate/grate-texture.c:
```diff
@@ -184,3 +268,216 @@ void grate_texture_clear(struct grate *grate, struct grate_texture *tex,
 		grate_error("host1x_gr2d_clear() failed: %d\n", err);
 }
+
+static unsigned log2_size(unsigned size)
```
We already have a function in grate-3d.c called `log2_size` that has slightly different semantics. I find this a bit confusing, perhaps we could use a different name for one of the, or use this version for both and check `log2_width == pixbuf->width && log2_height == pixbug->height` instead of `log2_width >= 0 && log2_height >= 0` in `grate_3d_set_texture_desc`?

digetx commented:
Consolidated log2_size() and switched to clz.

digetx commented on src/libgrate/grate-3d.h:
```diff
@@ -3,6 +3,8 @@
 
 #include <stdint.h>
 
+#define log2_size(s)		(31 - __builtin_clz(s))
```
Here it is.

kusma reviewed (approved):
Nice!

digetx commented:
You are too fast! I was going to write that I shouldn't have removed the grate_texture_set_max_lod() :)

Will re-add it.

kusma commented:
I didn't pull yet!

digetx commented:
Ah nice :) Will fix now then.

digetx committed:
grate: add grate_create_texture2()

That one infers the pixbuf size from the loaded image. Will be used by grate-font.

digetx commented:
Factored out grate_create_texture2() addition to a standalone commit and added test4_max_lod() to texture-filter test. Should be good now :)

kusma commented:
Great stuff :)
